core/domain/question_domain.py

```diff
@@ -0,0 +1,215 @@
+            self.property_name = change_dict['property_name']
+            self.new_value = change_dict['new_value']
+            self.old_value = change_dict.get('old_value')
+
+
+class Question(object):
+    """Domain object for a questions.
+
+    Attributes:
+        question_id: str. The unique ID of the question.
+        title: str. The title of the question.
+        question_data: dict. A dict representing the question data.
+        question_data_schema_version: int. The schema version for the data.
+        collection_id: str. The ID of the collection containing the question.
+        language_code: str. The ISO 639-1 code for the language this
+            question is written in.
+    """
+
+    def __init__(self, question_id, title, question_data,
+                 question_data_schema_version, collection_id, language_code):
+        """Constructs a Question domain object.
+
+        Args:
+            question_id: str. The unique ID of the question.
+            title: str. The title of the question.
+            question_data: dict. A dict representing the question data.
+            question_data_schema_version: int. The schema version for the data.
+            collection_id: str. The ID of the collection containing the
+                question.
+            language_code: str. The ISO 639-1 code for the language this
+                question is written in.
+        """
+        self.question_id = question_id
+        self.title = title
+        self.question_data = question_data
+        self.question_data_schema_version = question_data_schema_version
+        self.collection_id = collection_id
+        self.language_code = language_code
+
+    def to_dict(self):
+        """Returns a dict representing this Question domain object.
+
+        Returns:
+            dict. A dict representation of the Question instance.
+        """
+        return {
+            'question_id': self.question_id,
+            'title': self.title,
+            'question_data': self.question_data,
+            'question_data_schema_version': self.question_data_schema_version,
+            'collection_id': self.collection_id,
+            'language_code': self.language_code
+        }
+
+    def validate(self):
+        """Validates the Question domain object before it is saved."""
+
+        if not isinstance(self.question_id, basestring):
+            raise utils.ValidationError(
+                'Expected ID to be a string, received %s' % self.question_id)
+
+        if not isinstance(self.title, basestring):
+            raise utils.ValidationError(
+                'Expected title to be a string, received %s' % self.title)
+
+        if not isinstance(self.question_data, dict):
+            raise utils.ValidationError(
+                'Expected question_data to be a dict, received %s' %
+                self.question_data)
+        question_data = exp_domain.State.from_dict(self.question_data)
+        question_data.validate('', True)
```

Comment:
Shouldn't the first arg be {} or None? Perhaps None (as a sentinel value to indicate that the state belongs to a Question), but then the docstring for State.validate() should be updated accordingly and tests added.

Reply:
Updated.